### PR TITLE
Improve cluster creation metric

### PR DESCRIPTION
**Pull request by sjkingo**
This is a WIP from #91.

### CODE REVIEW

File: pkg/frontend/openshiftcluster_putorpatch_test.go [hunk 1]
```diff
@@ -138,6 +139,9 @@ func TestPutOrPatchOpenShiftCluster(t *testing.T) {
 							},
 						},
 					},
+					CorrelationData: &api.CorrelationData{
+						StartTime: time.Now(),
+					},
```
Comment (sjkingo) [inline pkg/frontend/openshiftcluster_putorpatch_test.go:144]: I'm not sure exactly what to put here to mock?

File: pkg/frontend/openshiftcluster_putorpatch_test.go [hunk 2]
```diff
@@ -138,6 +139,9 @@ func TestPutOrPatchOpenShiftCluster(t *testing.T) {
...
 				}
 
 				openShiftClusters.EXPECT().
```
Comment (sjkingo) [inline pkg/frontend/openshiftcluster_putorpatch_test.go:147]: Around here I expect to validate the timestamp is saved in the database.

File: pkg/backend/openshiftcluster.go [hunk 3]
```diff
@@ -106,6 +106,11 @@ func (ocb *openShiftClusterBackend) handle(ctx context.Context, log *logrus.Entr
 			return ocb.endLease(ctx, log, stop, doc, api.ProvisioningStateFailed, err)
 		}
 		if doc.OpenShiftCluster.Properties.Install == nil {
+			// emit the creation duration
+			duration := time.Now().Sub(doc.CorrelationData.StartTime).Seconds()
+			ocb.m.EmitGauge("backend.openshiftcluster.duration", int64(duration), map[string]string{
```
Comment (sjkingo) [inline pkg/backend/openshiftcluster.go:111]: Should this be called something else, or do we remove the other `backend.openshiftcluster.duration` emits?

File: pkg/api/correlation.go [hunk 4]
```diff
@@ -15,4 +17,6 @@ type CorrelationData struct {
 
 	// RequestID contains value of x-ms-request-id
 	RequestID string `json:"requestID,omitempty"`
+
+	StartTime time.Time `json:"startTime,omitempty"`
```
Comment (jim-minter) [inline pkg/api/correlation.go:21]: EnqueueTime or RequestTime may be better names

File: pkg/backend/openshiftcluster.go (hunk 3 above)
Comment (jim-minter) [reply pkg/backend/openshiftcluster.go:111]: wrong location.  The existing location in pkg/backend/openshiftcluster.go is about right, but it needs to detect that ProvisioningState has changed and only emit in that case.

File: pkg/frontend/openshiftcluster_putorpatch_test.go (hunk 1 above)
Comment (jim-minter) [reply pkg/frontend/openshiftcluster_putorpatch_test.go:144]: pick a value out of thin air. e.g. `time.Date(2000, 1, 1, 0, 0, 0, 0, time.UTC)`.

File: pkg/frontend/openshiftcluster_putorpatch.go [hunk 5]
```diff
@@ -46,6 +46,7 @@ func (f *frontend) putOrPatchOpenShiftCluster(w http.ResponseWriter, r *http.Req
 func (f *frontend) _putOrPatchOpenShiftCluster(ctx context.Context, r *http.Request, header *http.Header, converter api.OpenShiftClusterConverter, staticValidator api.OpenShiftClusterStaticValidator) ([]byte, error) {
 	body := r.Context().Value(middleware.ContextKeyBody).([]byte)
 	correlationData := r.Context().Value(middleware.ContextKeyCorrelationData).(*api.CorrelationData)
+	correlationData.RequestTime = f.now()
```
Comment (jim-minter) [inline pkg/frontend/openshiftcluster_putorpatch.go:49]: needs to be done on delete as well

File: pkg/backend/openshiftcluster.go [hunk 6]
```diff
@@ -55,7 +55,8 @@ func (ocb *openShiftClusterBackend) try(ctx context.Context) (bool, error) {
 			ocb.m.EmitGauge("backend.openshiftcluster.workers.count", int64(atomic.LoadInt32(&ocb.workers)), nil)
 			ocb.cond.Signal()
 
-			ocb.m.EmitGauge("backend.openshiftcluster.duration", time.Now().Sub(t).Milliseconds(), map[string]string{
+			duration := time.Now().Sub(doc.CorrelationData.RequestTime).Seconds()
```
Comment (jim-minter) [inline pkg/backend/openshiftcluster.go:58]: this code needs to detect if a change in provisioningstate has taken place, and only if it has log the duration time.

File: pkg/backend/openshiftcluster.go [hunk 7]
```diff
@@ -55,9 +57,15 @@ func (ocb *openShiftClusterBackend) try(ctx context.Context) (bool, error) {
...
 
-			ocb.m.EmitGauge("backend.openshiftcluster.duration", time.Now().Sub(t).Milliseconds(), map[string]string{
-				"state": string(doc.OpenShiftCluster.Properties.ProvisioningState),
-			})
+			// Emit duration metric for how long a create or delete takes to complete
+			if dequeuedState == api.ProvisioningStateCreating || dequeuedState == api.ProvisioningStateDeleting {
+				if ps := doc.OpenShiftCluster.Properties.ProvisioningState; ps == api.ProvisioningStateSucceeded || ps == api.ProvisioningStateFailed {
+					duration := ocb.now().Sub(doc.CorrelationData.RequestTime).Seconds()
+					ocb.m.EmitGauge("backend.openshiftcluster.duration", int64(duration), map[string]string{
+						"state": string(ps),
```
Comment (mjudeikis) [inline pkg/backend/openshiftcluster.go:65]: Am I missing something but how do we know this metric was for Delete or Create? We set the state of the end result, but not "what was the intended action". And I somehow think this should 2 separate metrics for created and deletes as this is 2 separate intentions and code paths.

Comment (sjkingo) [reply pkg/backend/openshiftcluster.go:65]: @jim-minter do you agree with mj that we should emit durations with the action in them, e.g.:

```
backend.openshiftcluster.<action>.duration
```

**Comment by sjkingo** [comment]
Currently working on unit test for emit duration

**Comment by sjkingo** [comment]
Rebased & squashed commits.

Ready for review.

**Pull request merged**